[FLINK-18064][docs] Added unaligned checkpointing to docs. (#12722)
apache/flink: docs/concepts/stateful-stream-processing.md

snapshots, we use the words *snapshot* and *checkpoint* interchangeably. Often
we also use the term *snapshot* to mean either *checkpoint* or *savepoint*.

> ### Aligned Checkpointing

Reviewer: Hum, not sure I'd change this title, it might be more confusing than helpful. The sentence you have below is clarifying enough ("Since Flink 1.11, checkpoints can be taken with or without alignment. In the 
following, we describe aligned checkpoints first.").

> The central part of Flink's fault tolerance mechanism is drawing consistent

Reviewer 2: To my understanding, there are were papers: one by Lamport and one by [DataArtisans](https://arxiv.org/pdf/1506.08603.pdf).
One of the differences is that the former proposed to persist channel state, while the latter proposed alignment (I might be wrong).
So it's probably better to move reference to the paper by Lamport to Unaligned section and mention "Flink" paper in this Aligned section.

Author: Good question. I briefly looked at it and it seems as if you are right. I pinged @StephanEwen for clarification and he responded that it is indeed closer, but there are still some difference:

> Flink checkpoints start at sources, propagate through a DAG and persist in-flight as needed
Chandy-Lamport assumes no DAG, starts everywhere at the same time (think: RPC goes to every operators) and all operators log until they saw all markers.

I'd probably leave the reference as is and add a clarifying line to unaligned section.

asynchronously. The checkpoint barriers don't travel in lock step and
operations can asynchronously snapshot their state.

Since Flink 1.11, checkpoints can be taken with or without alignment. In the 

> following, we describe aligned checkpoints first.

Reviewer 3: following section?

See [Restart Strategies]({% link dev/task_failure_recovery.md
%}#restart-strategies) for more information.

> ### Unaligned Checkpointing

Reviewer 2: Should we mention that this is an experimental feature?
I think it should be a separate statement in the end of the section.

Author: My understanding was that this document is rather an expanded glossary and talks about the concept and not the implementation. Thus, I'd leave the implementation state out of this place. The ops link will directly say that it's experimental in 1.11.

Reviewer 2: The page already goes quite deep into the details so I don't see why it shouldn't be mentioned here. Some users could benefit by ruling out the feature earlier if they are considering Flink or its configuration.

Author: Since we decided at a different point to drop the experimental label, I'd leave this section as is.

Starting with Flink 1.11, checkpointing can also be performed unaligned.
The basic idea is that checkpoints can overtake all in-flight data as long as 
the in-flight data becomes part of the operator state.

<div style="text-align: center">
  <img src="{% link fig/stream_unaligning.svg %}" alt="Unaligned checkpointing" style="width:100%; padding-top:10px; padding-bottom:10px;" />
</div>

> The figure depicts how an operator handle unaligned checkpoint barriers:

Reviewer: ```suggestion
The figure depicts how an operator handles unaligned checkpoint barriers:
```

> - The operator reacts on the first barrier that is stored in input buffers.

Reviewer 2: ```suggestion
- The operator reacts on the first barrier that is stored in its input buffers.
```

- It immediately forwards the barrier to the downstream operator by adding it 
  to the end of the output buffers.

> - The operator marks all overtaken buffers, which are asynchronously stored in
>   the state backend together with the other operator state.

Reviewer 2: ```suggestion
- The operator marks all overtaken records to be stored asynchronously and creates a snapshot of its own state.
```

> - It only briefly stops the processing of input to mark the buffers, forward 
>   the barrier and create the snapshot of the other state.

Reviewer 2: It reads like a separate step (it's not, right?)
```suggestion
Compared to aligned checkpoints, the operator doesn't need to suspend any of its inputs.
```

Author: Yes, I have removed the bullet and made it a summarizing sentence.

  
Unaligned checkpointing ensures that barriers are arriving at the sink as fast 

> as possible. It's especially suited for application with at least one slow 

Reviewer: ```suggestion
as possible. It's especially suited for applications with at least one slow 
```

moving data path, where alignment times can reach hours. However, since it's
adding additional I/O pressure to state backends, it doesn't help when the I/O
to the state backends is the bottleneck. See the more in-depth discussion in 
[ops]({% link ops/state/checkpoints.md %}#unaligned-checkpoints)
for other limitations.

> Note that savepoints will always be aligned.

Reviewer 2: :+1:

](http://research.microsoft.com/en-us/um/people/lamport/pubs/chandy.pdf), but
Flink still inserts the barrier in the sources to avoid overloading the
checkpoint coordinator.

<div style="text-align: center">
  <img src="{% link fig/stream_unaligning.svg %}" alt="Unaligned checkpointing" style="width:100%; padding-top:10px; padding-bottom:10px;" />
</div>

The figure depicts how an operator handles unaligned checkpoint barriers:

- The operator reacts on the first barrier that is stored in its input buffers.
- It immediately forwards the barrier to the downstream operator by adding it 
  to the end of the output buffers.
- The operator marks all overtaken records to be stored asynchronously and 
  creates a snapshot of its own state.
 
Consequently, the operator only briefly stops the processing of input to mark
the buffers, forwards the barrier, and creates the snapshot of the other state.
  
Unaligned checkpointing ensures that barriers are arriving at the sink as fast 
as possible. It's especially suited for applications with at least one slow 
moving data path, where alignment times can reach hours. However, since it's

> adding additional I/O pressure to state backends, it doesn't help when the I/O

Reviewer 3: `I/O pressure to state backends` -> ` I/O pressure`, as it's not using state backends per se.
